crates/ruff_cli/src/lib.rs

```diff
@@ -171,12 +172,26 @@ fn check(args: CheckArgs, log_level: LogLevel) -> Result<ExitStatus> {
         cli.stdin_filename.as_deref(),
     )?;
 
+    let mut writer: Box<dyn Write> = match cli.output_file {
+        Some(path) if !cli.watch => {
+            colored::control::set_override(false);
+            let file = File::create(path)?;
+            Box::new(BufWriter::new(file))
+        }
+        _ => Box::new(BufWriter::new(io::stdout().lock())),
```

Comment:
The output is now locked significantly longer than before. This can lead to unexpected behavior, e.g using `dbg` during this time won't work. 

We also need to be careful to drop this `writer` at the right time (e.g. only when entering the watch). 

What are your thoughts on instead defining our own `enum` that implements `Write` that has a `lock` operation (which is a no-op for files)?

Reply:
Good point, I missed that. Defining our own `enum` is a good idea although I'm unable to understand what would the `lock` operation be like. Would it be an additional method implemented on the said `enum`? Acquiring a lock on every write seems unnecessary.

Reply:
The idea would be to have aon object similar to stout that has a lock object that then aquires the lock and releases it on drop. 

I can try to find the source in Rome tomorrow. I believe it used something similar

Reply:
Where did we land on this @dhruvmanila?

Reply:
Sorry for letting this sit, I'll get on it today.

As per our benchmarks the lock is giving us performance benefits so will be moving with Micha's idea.